crates/services/p2p/src/peer_report.rs

```diff
@@ -0,0 +1,452 @@
+        peer_id: PeerId,
+    },
+    PeerIdentified {
+        peer_id: PeerId,
+        agent_version: String,
+        addresses: Vec<Multiaddr>,
+    },
+    PeerInfoUpdated {
+        peer_id: PeerId,
+        block_height: BlockHeight,
+    },
+    /// Informs p2p service / PeerManager to check health of reserved nodes' connections
+    CheckReservedNodesHealth,
+    /// Informs p2p service / PeerManager to perform reputation decay of connected nodes
+    PerformDecay,
+}
+
+// `Behaviour` that reports events about peers
+pub struct PeerReportBehaviour {
+    heartbeat: Heartbeat,
+    identify: Identify,
+    pending_events: VecDeque<PeerReportEvent>,
+    // regulary checks if reserved nodes are connected
+    health_check: Interval,
+    decay_interval: Interval,
+}
+
+impl PeerReportBehaviour {
+    pub(crate) fn new(config: &Config) -> Self {
+        let identify = {
+            let identify_config =
+                IdentifyConfig::new("/fuel/1.0".to_string(), config.keypair.public());
+            if let Some(interval) = config.identify_interval {
+                Identify::new(identify_config.with_interval(interval))
+            } else {
+                Identify::new(identify_config)
+            }
+        };
+
+        let heartbeat =
+            Heartbeat::new(config.heartbeat_config.clone(), BlockHeight::default());
+
+        Self {
+            heartbeat,
+            identify,
+            pending_events: VecDeque::default(),
+            health_check: time::interval(Duration::from_secs(
+                HEALTH_CHECK_INTERVAL_IN_SECONDS,
+            )),
+            decay_interval: time::interval(Duration::from_secs(
+                REPUTATION_DECAY_INTERVAL_IN_SECONDS,
+            )),
+        }
+    }
+
+    pub fn update_block_height(&mut self, block_height: BlockHeight) {
+        self.heartbeat.update_block_height(block_height);
+    }
+}
+
+impl NetworkBehaviour for PeerReportBehaviour {
+    type ConnectionHandler = IntoConnectionHandlerSelect<
+        <Heartbeat as NetworkBehaviour>::ConnectionHandler,
+        <Identify as NetworkBehaviour>::ConnectionHandler,
+    >;
+    type OutEvent = PeerReportEvent;
+
+    fn new_handler(&mut self) -> Self::ConnectionHandler {
+        IntoConnectionHandler::select(
+            self.heartbeat.new_handler(),
+            self.identify.new_handler(),
+        )
+    }
+
+    fn addresses_of_peer(&mut self, peer_id: &PeerId) -> Vec<Multiaddr> {
+        self.identify.addresses_of_peer(peer_id)
+    }
+
+    fn on_swarm_event(&mut self, event: FromSwarm<Self::ConnectionHandler>) {
+        match event {
+            FromSwarm::ConnectionEstablished(connection_established) => {
+                let ConnectionEstablished {
+                    peer_id,
+                    other_established,
+                    ..
+                } = connection_established;
+
+                self.heartbeat
+                    .on_swarm_event(FromSwarm::ConnectionEstablished(
```

Comment:
Most of these functions are just forwarding events. A helper function could reduce duplication

Reply:
All of these events need to be extracted separately and passed down to two different behaviours.

I don't see a way to reduce "duplication" since all events are different, this is caused because parent `ConnectionHandler` is different than the children behaviours'. So it first needs to be extracted and sent down separately to the children.

Reply:
Made a draft PR here which extracts these conversions out into separate functions: https://github.com/FuelLabs/fuel-core/pull/1109

For the events that are simply pass-throughs, I wonder if the helper methods like `.map_handler` and `.map_out` could be used 🤔

Reply:
That looks good, but I understood that freesig meant de-duplicating forwarding of events in `on_swarm_event`

Reply:
@Voxelot here's my take at the `on_swarm_event` de-dup